File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char,int> last;
        int start = 1, ans = 0;
        for(int i = 0; i < s.length(); i++) {
            if(last[s[i]] != 0) {
                start = max(last[s[i]] + 1,start);
            }
            last[s[i]] = i + 1;
            ans = max(ans,i + 1 - start + 1);
        }
        return ans;
    }
};

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        
        # This dictionary stores the position where we find a character
        lastOccuredAt = dict()
        
        # Stores the index from where the current substring starts from
        currSubstringStartsAt = 0
        
        # Stores the final answer
        ans = 0
        
        # Looping through the string once
        for i in range(len(s)):
            
            currChar = s[i]
            
            # Checks if current character has previously occured 
            # if it hasn't just take in current substring
            # if it has check if the index at which it occured is before the start index for curr substring
            if (currChar not in lastOccuredAt or lastOccuredAt[currChar] < currSubstringStartsAt) :
                lastOccuredAt[currChar] = i
                ans = max(ans, i - currSubstringStartsAt + 1)
                
            else:
                currSubstringStartsAt = lastOccuredAt[currChar] + 1
                lastOccuredAt[currChar] = i
                ans = max(ans, i - currSubstringStartsAt + 1)
        
        return ans
        # Time Complexity: O(n)  Space Complexity: O(k) where k is number of distinct chars in a string